File: src/CSharp/200.NumberOfIslands.cs

using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.NumberOfIslands
{
public class Solution {
    public void dfs(char[][] grid, int r, int c) {
        int nr = grid.Length;
        int nc = grid[0].Length;
        if (r < 0 || c < 0 || r >= nr || c >= nc || grid[r][c] == '0') {
            return;
        }
        grid[r][c] = '0';
        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
    public int NumIslands(char[][] grid) {
        if (grid == null || grid.Length == 0) {
            return 0;
        }
        int nr = grid.Length;
        int nc = grid[0].Length;
        int num_islands = 0;
        for (int r = 0; r < nr; ++r) {
            for (int c = 0; c < nc; ++c) {
                if (grid[r][c] == '1') {
                    ++num_islands;
                    dfs(grid, r, c);
                }
            }
        }
        return num_islands;
    }
}
}

File: src/Python/101-200/200.NumberOfIslands.py

class Solution:
    def dfs(self, grid, r, c):
        nr = len(grid)
        nc = len(grid[0])
        grid[r][c] = '0'
        if r - 1 >= 0 and grid[r-1][c] == '1': 
            self.dfs(grid, r - 1, c)
        if r + 1 < nr and grid[r+1][c] == '1': 
            self.dfs(grid, r + 1, c)
        if c - 1 >= 0 and grid[r][c-1] == '1': 
            self.dfs(grid, r, c - 1)
        if c + 1 < nc and grid[r][c+1] == '1': 
            self.dfs(grid, r, c + 1)

    def numIslands(self, grid: List[List[str]]) -> int:
        nr = len(grid)
        if nr == 0:
            return 0
        nc = len(grid[0])
        num_islands = 0
        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == '1':
                    num_islands += 1
                    self.dfs(grid, r, c)
        return num_islands

if __name__ == "__main__":
    solution = Solution()


